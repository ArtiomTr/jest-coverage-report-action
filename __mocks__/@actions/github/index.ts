export const context: Record<string, unknown> = {};

export const mockContext = (newContext: Record<string, unknown>) => {
    Object.assign(context, newContext);
};

export const clearContextMock = () => {
    Object.keys(context).map((key) => {
        delete context[key];
    });
};

export const getOctokit = jest.fn();
