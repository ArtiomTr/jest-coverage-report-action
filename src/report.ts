import core from '@actions/core';
import { ZodError } from 'zod';

import { getCurrentBranch } from './git';
import { getOptions } from './options';

export interface InvalidOptionsReport {
    type: 'invalid-options';
    error: ZodError;
}

export type CovbotReport = InvalidOptionsReport;

export const getReport = async (): Promise<CovbotReport> => {
    core.debug('[stage]: getting options');
    const options = getOptions();

    /**
     * Report error if failed to parse options.
     * There is nothing more we could do.
     */
    if (!options.success) {
        return {
            type: 'invalid-options',
            error: options.error,
        };
    }

    /**
     * Try to get current branch on current repository.
     * This is needed just for consistency - when action ends work, return to initial branch.
     * It will automatically report warning, if getting this branch failed.
     */
    core.debug('[stage]: getting current branch');
    const currentBranch = await getCurrentBranch();

    /**
     * If it is pull request, or pull request number is present, get PR info
     */
    const pullRequest = getPullRequestInfo();
};
