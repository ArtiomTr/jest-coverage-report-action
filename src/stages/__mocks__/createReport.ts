export const createReport = jest.fn();
