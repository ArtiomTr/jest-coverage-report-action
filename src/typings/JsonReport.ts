export type JsonReport = {
    numFailedTestSuites: number;
    numFailedTests: number;
    numPassedTestSuites: number;
    numPassedTests: number;
    numPendingTestSuites: number;
    numPendingTests: number;
    numRuntimeErrorTestSuites: number;
    numTodoTests: number;
    numTotalTestSuites: number;
    numTotalTests: number;
    openHandles?: unknown[];
    snapshot: Snapshot;
    startTime: number;
    success: boolean;
    testResults?: TestResult[];
    wasInterrupted: boolean;
    coverageMap: CoverageMap;
};

export type Snapshot = {
    added: number;
    didUpdate: boolean;
    failure: boolean;
    filesAdded: number;
    filesRemoved: number;
    filesRemovedList?: unknown[];
    filesUnmatched: number;
    filesUpdated: number;
    matched: number;
    total: number;
    unchecked: number;
    uncheckedKeysByFile?: unknown[];
    unmatched: number;
    updated: number;
};

export type TestResult = {
    assertionResults?: AssertionResult[];
    endTime: number;
    message: string;
    name: string;
    startTime: number;
    status: string;
    summary: string;
};

export type AssertionResult = {
    ancestorTitles?: string[];
    failureMessages?: string[];
    fullName: string;
    location: Location;
    status: string;
    title: string;
};

export type Location = {
    column?: number;
    line: number;
};

export type Range = {
    start?: Location;
    end?: Location;
};

export type CoverageMap = Record<string, FileCoverage>;

export type FileCoverage = {
    path: string;
    statementMap: StatementMap;
    fnMap: FunctionMap;
    branchMap: BranchMap;
    s: HitMap;
    f: HitMap;
    b: ArrayHitMap;
};

export type StatementMap = Record<number, StatementCoverage>;

export type StatementCoverage = {
    start: Location;
    end: Location;
};

export type FunctionMap = Record<number, FunctionCoverage>;

export type FunctionCoverage = {
    name: string;
    decl: Range;
    loc: Range;
};

export type BranchMap = Record<number, BranchCoverage>;

export type BranchCoverage = {
    loc: Range;
    type: string;
    locations?: Range[];
};

export type HitMap = Record<number, number>;

export type ArrayHitMap = Record<number, number[]>;
