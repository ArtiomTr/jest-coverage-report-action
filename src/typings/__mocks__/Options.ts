export const getOptions = jest.fn(() => ({
    token: 'TOKEN',
    testScript: 'test script',
    iconType: 'emoji',
    annotations: 'all',
    threshold: 80,
    packageManager: 'npm',
    skipStep: 'none',
}));
