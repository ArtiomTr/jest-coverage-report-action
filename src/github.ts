import { context, getOctokit } from '@actions/github';

import {
    PullRequestInfo,
    PullRequestInfoQuery,
    PullRequestInfoQueryVariables,
} from './generated/graphql';

export type Octokit = ReturnType<typeof getOctokit>;

function graphql<TQuery, TQueryVariables extends Record<string, unknown>>(
    octokit: Octokit,
    source: string,
    variables: TQueryVariables
) {
    return octokit.graphql<TQuery>(source, {
        ...variables,
    });
}

export const getPullRequestInfo = async (
    octokit: Octokit,
    prNumber?: number
) => {
    if (context?.payload?.pull_request) {
        return context?.payload?.pull_request;
    }

    if (!prNumber) {
        return;
    }

    const output = await graphql<
        PullRequestInfoQuery,
        PullRequestInfoQueryVariables
    >(octokit, PullRequestInfo, {
        repo: context.repo.repo,
        owner: context.repo.owner,
        prNumber,
    });

    output.repository?.pullRequest?.baseRef?.repository.nameWithOwner;
};
