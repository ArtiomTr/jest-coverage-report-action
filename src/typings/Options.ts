import { getInput } from '@actions/core';
import { context, getOctokit } from '@actions/github';
import * as yup from 'yup';

import { icons } from '../format/strings.json';

export type IconType = keyof typeof icons;

export type AnnotationType = 'all' | 'none' | 'coverage' | 'failed-tests';
export type PackageManagerType = 'npm' | 'yarn' | 'pnpm';
export type SkipStepType = 'all' | 'none' | 'install';
export type PullRequest = {
    base: { ref: string };
    head: { ref: string; sha: string };
    number: number;
};
export type Options = {
    token: string;
    testScript: string;
    iconType: IconType;
    annotations: AnnotationType;
    threshold?: number;
    workingDirectory?: string;
    packageManager: PackageManagerType;
    skipStep: SkipStepType;
    customTitle?: string;
    coverageFile?: string;
    baseCoverageFile?: string;
    prNumber: null | number;
    pullRequest: null | PullRequest;
};

const validAnnotationOptions: Array<AnnotationType> = [
    'all',
    'none',
    'coverage',
    'failed-tests',
];

const packageManagerOptions: Array<PackageManagerType> = [
    'npm',
    'yarn',
    'pnpm',
];

const validIconOptions = Object.keys(icons);

const validSkipStepOptions: Array<SkipStepType> = ['all', 'none', 'install'];

const optionSchema = yup.object().shape({
    token: yup.string().required(),
    testScript: yup.string().required(),
    iconType: yup.string().required().oneOf(validIconOptions),
    annotations: yup.string().required().oneOf(validAnnotationOptions),
    threshold: yup
        .number()
        .transform((value) => (isNaN(value) ? undefined : value))
        .min(0)
        .max(100),
    workingDirectory: yup.string(),
    packageManager: yup.string().required().oneOf(packageManagerOptions),
    skipStep: yup.string().required().oneOf(validSkipStepOptions),
    customTitle: yup.string(),
    coverageFile: yup.string(),
    baseCoverageFile: yup.string(),
    prNumber: yup.number().nullable(),
    pullRequest: yup.object().nullable(),
});

export const shouldInstallDeps = (skipStep: SkipStepType): Boolean =>
    !['all', 'install'].includes(skipStep);

export const shouldRunTestScript = (skipStep: SkipStepType): Boolean =>
    !['all'].includes(skipStep);

export const getOptions = async (): Promise<Options> => {
    const token = getInput('github-token', {
        required: true,
    });
    const octokit = getOctokit(token);
    const testScript = getInput('test-script');
    const threshold = getInput('threshold');
    const workingDirectory = getInput('working-directory');
    const iconType = getInput('icons');
    const annotations = getInput('annotations');
    const packageManager = getInput('package-manager');
    const skipStep = getInput('skip-step');
    const customTitle = getInput('custom-title');
    const coverageFile = getInput('coverage-file');
    const baseCoverageFile = getInput('base-coverage-file');
    const prNumber: number | null = Number(
        getInput('prnumber') || context?.payload?.pull_request?.number
    );
    let pullRequest = context?.payload?.pull_request || null;

    if (!pullRequest && !Number.isNaN(prNumber)) {
        const { data: pr } = await octokit.pulls.get({
            owner: context.repo.owner,
            repo: context.repo.repo,
            pull_number: prNumber,
        });
        pullRequest = pr as PullRequest;
    }

    try {
        const options: Options = (await optionSchema.validate({
            token,
            testScript,
            threshold,
            workingDirectory,
            iconType,
            annotations,
            packageManager,
            skipStep,
            customTitle,
            coverageFile,
            baseCoverageFile,
            prNumber: prNumber || null,
            pullRequest,
        })) as Options;

        return options;
    } catch (err) {
        if (err instanceof yup.ValidationError) {
            throw new Error(
                [err.message, ...err.errors].filter(Boolean).join('\n')
            );
        }

        throw err;
    }
};
