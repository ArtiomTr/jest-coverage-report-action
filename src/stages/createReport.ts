import { context } from '@actions/github';

import { getReportTag } from '../constants/getReportTag';
import { GITHUB_MESSAGE_SIZE_LIMIT } from '../constants/GITHUB_MESSAGE_SIZE_LIMIT';
import { formatCoverage } from '../format/formatCoverage';
import { formatErrors } from '../format/formatErrors';
import { formatRunReport } from '../format/formatRunReport';
import { formatThresholdResults } from '../format/formatThresholdResults';
import { getFailureDetails } from '../format/getFailureDetails';
import { getTestRunSummary } from '../format/summary/getTestRunSummary';
import template from '../format/template.md';
import { JsonReport } from '../typings/JsonReport';
import { Options } from '../typings/Options';
import { SummaryReport, TestRunReport } from '../typings/Report';
import { ThresholdResult } from '../typings/ThresholdResult';
import { DataCollector } from '../utils/DataCollector';
import { i18n } from '../utils/i18n';
import { insertArgs } from '../utils/insertArgs';

export const getSha = () =>
    context.payload.after ??
    context.payload.pull_request?.head.sha ??
    context.sha;

export const createReport = (
    dataCollector: DataCollector<JsonReport>,
    options: Options,
    thresholdResults: ThresholdResult[]
): SummaryReport => {
    const { workingDirectory, customTitle } = options;

    const { errors, data } = dataCollector.get();
    const [headReport, baseReport] = data;
    const formattedErrors = formatErrors(errors);

    const formattedThresholdResults = formatThresholdResults(thresholdResults);
    const coverage = formatCoverage(headReport, baseReport, undefined, false);
    const runReport: TestRunReport = {
        title: i18n(headReport.success ? 'testsSuccess' : 'testsFail'),
        summary: getTestRunSummary(headReport),
        failures: getFailureDetails(headReport),
    };
    const formattedReport = formatRunReport(runReport);

    let templateText = insertArgs(template, {
        body: [
            formattedErrors,
            formattedThresholdResults,
            coverage,
            formattedReport,
        ].join('\n'),
        dir: workingDirectory || '',
        tag: getReportTag(options),
        title: insertArgs(customTitle || i18n('summaryTitle'), {
            dir: workingDirectory ? `for \`${workingDirectory}\`` : '',
        }),
        sha: getSha(),
    });

    if (templateText.length > GITHUB_MESSAGE_SIZE_LIMIT) {
        const reducedCoverage = formatCoverage(
            headReport,
            baseReport,
            undefined,
            true
        );

        templateText = insertArgs(template, {
            body: [
                formattedErrors,
                formattedThresholdResults,
                reducedCoverage,
                formattedReport,
            ].join('\n'),
            dir: workingDirectory || '',
            tag: getReportTag(options),
            title: insertArgs(customTitle || i18n('summaryTitle'), {
                dir: workingDirectory ? `for \`${workingDirectory}\`` : '',
            }),
            sha: getSha(),
        });
    }

    return {
        text: templateText,
        runReport,
    };
};
