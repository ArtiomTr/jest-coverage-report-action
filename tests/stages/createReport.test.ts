import * as all from '@actions/github';

import { createReport, getSha } from '../../src/stages/createReport';
import { JsonReport } from '../../src/typings/JsonReport';
import { Options } from '../../src/typings/Options';
import { createDataCollector } from '../../src/utils/DataCollector';
import report from '../mock-data/jsonReport.json';

const { mockContext, clearContextMock } = all as any;

const DEFAULT_OPTIONS: Options = {
    token: '',
    testScript: '',
    iconType: 'emoji',
    annotations: 'all',
    packageManager: 'npm',
    skipStep: 'all',
    prNumber: 5,
    pullRequest: {
        number: 5,
        head: { sha: '123456', ref: '123' },
        base: { ref: '456' },
    },
    output: ['comment'],
};

describe('createReport', () => {
    it('should match snapshots', async () => {
        const dataCollector = createDataCollector<JsonReport>();
        dataCollector.add(report);

        mockContext({ payload: { after: '123456' } });
        expect(
            await createReport(
                dataCollector,
                {
                    ...DEFAULT_OPTIONS,
                    workingDirectory: 'custom directory',
                },
                []
            )
        ).toMatchSnapshot();
        expect(
            await createReport(dataCollector, DEFAULT_OPTIONS, [])
        ).toMatchSnapshot();

        expect(
            await createReport(
                dataCollector,
                {
                    ...DEFAULT_OPTIONS,
                    workingDirectory: 'directory',
                    customTitle: 'Custom title with directory - {{ dir }}',
                },
                []
            )
        ).toMatchSnapshot();

        clearContextMock();
    });

    it('should match snapshots (failed report)', async () => {
        const dataCollector = createDataCollector<JsonReport>();
        dataCollector.add({ ...report, success: false });

        mockContext({ payload: { after: '123456' } });

        expect(
            await createReport(dataCollector, DEFAULT_OPTIONS, [])
        ).toMatchSnapshot();

        clearContextMock();
    });

    it('should extract commit shasum from context', async () => {
        mockContext({ payload: { after: '123456' } });
        expect(getSha()).toBe('123456');
        clearContextMock();

        mockContext({ payload: { pull_request: { head: { sha: '123456' } } } });
        expect(getSha()).toBe('123456');
        clearContextMock();

        mockContext({ payload: {}, sha: '123456' });
        expect(getSha()).toBe('123456');
        clearContextMock();
    });
});
