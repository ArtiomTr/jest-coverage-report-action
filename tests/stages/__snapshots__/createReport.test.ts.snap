// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createReport > should match snapshots (failed report) 1`] = `
{
  "runReport": {
    "failures": "",
    "summary": "Failed tests: 0/12. Failed suites: 0/5.",
    "title": "Test suite run failed",
  },
  "text": "/src/format/template.txt",
}
`;

exports[`createReport > should match snapshots 1`] = `
{
  "runReport": {
    "failures": "",
    "summary": "12 tests passing in 5 suites.",
    "title": "Test suite run success",
  },
  "text": "/src/format/template.txt",
}
`;

exports[`createReport > should match snapshots 2`] = `
{
  "runReport": {
    "failures": "",
    "summary": "12 tests passing in 5 suites.",
    "title": "Test suite run success",
  },
  "text": "/src/format/template.txt",
}
`;

exports[`createReport > should match snapshots 3`] = `
{
  "runReport": {
    "failures": "",
    "summary": "12 tests passing in 5 suites.",
    "title": "Test suite run success",
  },
  "text": "/src/format/template.txt",
}
`;
